packages/databyss-editor/state/reducer.ts

```diff
@@ -78,9 +86,103 @@ export default (
       const { payload } = action
 
       // default nextSelection to `payload.selection` (which may be undef)
-      const nextSelection = payload.selection
+      let nextSelection = payload?.selection
 
       switch (action.type) {
+        case CUT: {
+          deleteBlocksAtSelection({ state: state, draftState: draft })
+          break
+        }
+        case PASTE: {
+          const _frag = payload.data
+
+          // check if paste is occuring on an atomic block
+
+          const {anchor: _startAnchor} = sortSelection(state.selection)
+          const _startBlock = state.blocks[_startAnchor.index]
+
+          // if selection in the middle of atomic prevent paste
+          if(!isSelectionCollapsed(state.selection) || _startBlock.text.textValue.length !== _startAnchor.offset){            
```

Comment:
What about if it's at the beginning of the atomic? In that case, I think we should paste the fragment before the atomic...
In fact, in these cases where it's in the middle or end, we should paste the fragment after the atomic instead of preventing it.